common-yaml: YamlJsonParser.getObject()/getValue() corrupt parser position from a JSON-B custom deserializer or adapter

## Summary

`YamlJsonParser` (`runtime/common-yaml/src/main/java/io/aklivity/zilla/runtime/common/yaml/internal/json/YamlJsonParser.java`) is used as the `JsonProvider` behind `YamlJson.provider()`, which several `common-*` and `binding-*` modules pass to `JsonbBuilder` for JSON-B binding of YAML/JSON documents (e.g. `common-openapi`'s `OpenapiParser`).

When a custom `JsonbDeserializer<T>` or `JsonbAdapter<T, JsonObject>` is registered for a type and JSON-B needs to materialize the current value as a `JsonObject`/`JsonValue` to hand to that deserializer/adapter, the underlying call eventually reaches `YamlJsonParser.getObject()` / `getValue()`. In that specific call path, the parser's internal position tracking is corrupted. Two distinct failure modes were found, both apparently sharing one root cause:

1. **Loud failure — custom deserializer for a type used as a Map value.** A `JsonbDeserializer<T>` registered for a type `T` that appears as `Map<String, T>` (e.g. a security-scheme-like object keyed by name) throws when it calls `parser.getObject()`:
   ```
   jakarta.json.bind.JsonbException: Internal error: There are no more elements available!
       at org.eclipse.yasson.internal.DeserializationContextImpl.deserializeItem(...)
       ...
   Caused by: java.util.NoSuchElementException: There are no more elements available!
       at org.eclipse.yasson.internal.deserializer.YassonParser.validate(YassonParser.java:192)
       at org.eclipse.yasson.internal.deserializer.YassonParser.getObject(YassonParser.java:121)
       at <your>.SecuritySchemeDeserializer.deserialize(SecuritySchemeDeserializer.java:44)
       at org.eclipse.yasson.internal.deserializer.UserDefinedDeserializer.deserialize(...)
       at org.eclipse.yasson.internal.deserializer.MapDeserializer.deserializeValue(MapDeserializer.java:110)
       ...
   ```

2. **Silent failure — `JsonbAdapter<T, JsonObject>` for a plain field.** Registering `JsonbAdapter<T, JsonObject>` for a type `T` that is a *plain field* (not a Map value) does not throw at all, but silently corrupts the parser cursor: JSON properties that come *after* the adapted field in the source document fail to bind (the field comes back `null`), with zero exceptions raised and the adapter's own `adaptFromJson` sometimes not even invoked for the affected value. This is harder to detect than (1) because there is no stack trace — it just looks like a field failed to populate.

Both failures reproduce with `openapi: 3.1.0`-shaped documents; example minimal shape below.

## Repro

Given (via `YamlJson.provider()` + a `Jsonb` built with `.withDeserializers(...)`/`.withAdapters(...)`):
```yaml
openapi: 3.1.0
info:
  title: sample
  version: 1.0.0
paths:
  /items:
    get:
      operationId: listItems
      responses:
        '200':
          description: ok
components:
  securitySchemes:
    bearerAuth:
      type: http
      scheme: bearer
```

Register a `JsonbDeserializer<SecurityScheme>` (bound to the `bearerAuth` entry, i.e. dispatched as a `Map<String, SecurityScheme>` value) whose `deserialize()` calls `parser.getObject()` as its first action → throws `NoSuchElementException` as above.

Register a `JsonbAdapter<Operation, JsonObject>` instead (bound to the `get` field on a path item, a plain field, not a Map value) → no exception, but `components` (the sibling property after `paths` in the document) comes back `null` from the overall `fromJson()` call, with `errors.size() == 0`.

## What does *not* reproduce it

Manually driving the parser via only the primitive event API — `next()`/`getString()`/`getBigDecimal()`/`Event` — and hand-building the `JsonObject`/`JsonArray` tree via `Json.createObjectBuilder()`/`Json.createArrayBuilder()`, *without* ever calling `getObject()`, `getValue()`, or `getArray()` — works correctly in every context tested, including the two failing cases above. This suggests the bug is specific to `YamlJsonParser`'s implementation of those three convenience methods (or their interaction with Yasson's internal dispatch for custom deserializers/adapters/`ctx.deserialize(...)`), not the underlying event stream itself.

## Where this is currently worked around

`runtime/common-openapi/src/main/java/io/aklivity/zilla/runtime/common/openapi/model/OpenapiJsonValues.java` (added in PR #1994, branch `feature/openapi-generic-extensions`) implements the primitive-event-only materializer described above, used by `OpenapiOperationDeserializer.java` and `OpenapiSecuritySchemeDeserializer.java` instead of `parser.getObject()`. That PR needs this workaround to capture OpenAPI `x-*` specification-extension properties (arbitrary sibling keys JSON-B has no "any-setter" for) during a single parse pass.

## Ask

Investigate `YamlJsonParser.getObject()`/`getValue()`/`getArray()` (and whatever `YassonParser` delegates to when materializing a `JsonObject`/`JsonValue` for a custom deserializer, adapter, or `DeserializationContext.deserialize(JsonValue.class, parser)` call) to find why position tracking is lost specifically in these dispatch contexts, and fix it there so downstream modules don't need a parser-event-level workaround. Happy to discuss the best fix approach here or in a follow-up PR targeting `common-yaml` only — not proposing to bundle this into #1994.

--- a/runtime/common-yaml/src/main/java/io/aklivity/zilla/runtime/common/yaml/internal/json/YamlJsonParser.java
+++ b/runtime/common-yaml/src/main/java/io/aklivity/zilla/runtime/common/yaml/internal/json/YamlJsonParser.java
@@ -225,7 +225,15 @@ public final class YamlJsonParser implements JsonParser
         {
             throw new IllegalStateException("No value is available for current event");
         }
-        return value(new int[] {current});
+        int[] at = new int[] {current};
+        JsonValue value = value(at);
+        // per the JsonParser contract, materializing a structured value advances the parser to the
+        // matching END_OBJECT/END_ARRAY event; a scalar value is already positioned there, so this is a
+        // no-op for scalars. Skipping this step leaves cursor/current stale, so a subsequent next() replays
+        // the just-materialized value's events instead of continuing after it.
+        current = at[0] - 1;
+        cursor = at[0];
+        return value;
     }
 
     @Override

--- a/runtime/common-yaml/src/test/java/io/aklivity/zilla/runtime/common/yaml/YamlJsonParserTest.java
+++ b/runtime/common-yaml/src/test/java/io/aklivity/zilla/runtime/common/yaml/YamlJsonParserTest.java
@@ -804,18 +804,108 @@ class YamlJsonParserTest
         assertThrows(IllegalStateException.class, parser::getValue);
         assertEquals(START_OBJECT, parser.next());
         assertEquals(START_OBJECT, parser.currentEvent());
+        assertThrows(IllegalStateException.class, parser::skipArray);
         JsonObject object = parser.getObject();
         assertEquals("test", object.getString("name"));
         assertEquals(1, object.getJsonArray("values").getInt(0));
-        assertEquals(2, parser.getObjectStream().count());
-        assertEquals(1, parser.getValueStream().count());
-        assertThrows(IllegalStateException.class, parser::skipArray);
-
-        while (parser.hasNext())
-        {
-            parser.next();
-        }
+        // getObject() must advance the parser to the matching END_OBJECT, consuming the whole document
+        assertEquals(END_OBJECT, parser.currentEvent());
+        assertFalse(parser.hasNext());
         assertThrows(JsonParsingException.class, parser::next);
+    }
+
+    @Test
+    void shouldExposeObjectAndValueStreams()
+    {
+        JsonParser parser = parserFor("""
+            name: test
+            values: [1]
+            """);
+        assertEquals(START_OBJECT, parser.next());
+        assertEquals(2, parser.getObjectStream().count());
+        assertFalse(parser.hasNext());
+
+        parser = parserFor("""
+            name: test
+            """);
+        assertEquals(START_OBJECT, parser.next());
+        assertEquals(1, parser.getValueStream().count());
+        assertFalse(parser.hasNext());
+    }
+
+    @Test
+    void shouldAdvancePastMaterializedObjectSoSiblingPropertiesRemainReadable()
+    {
+        // reproduces the corruption from GH-1997: getObject()/getValue() must not leave the parser
+        // positioned back at the start of the value it just materialized, or the sibling properties that
+        // follow it in the document are misread or skipped entirely
+        JsonParser parser = parserFor("""
+            first:
+              nested: value
+            second: after
+            """);
+
+        assertEquals(START_OBJECT, parser.next());
+        assertEquals(KEY_NAME, parser.next());
+        assertEquals("first", parser.getString());
+        assertEquals(START_OBJECT, parser.next());
+
+        JsonObject nested = parser.getObject();
+        assertEquals("value", nested.getString("nested"));
+        assertEquals(END_OBJECT, parser.currentEvent());
+
+        assertEquals(KEY_NAME, parser.next());
+        assertEquals("second", parser.getString());
+        assertEquals(VALUE_STRING, parser.next());
+        assertEquals("after", parser.getString());
+        assertEquals(END_OBJECT, parser.next());
+        assertFalse(parser.hasNext());
+    }
+
+    @Test
+    void shouldAdvancePastMaterializedArraySoSiblingPropertiesRemainReadable()
+    {
+        JsonParser parser = parserFor("""
+            first: [1, 2]
+            second: after
+            """);
+
+        assertEquals(START_OBJECT, parser.next());
+        assertEquals(KEY_NAME, parser.next());
+        assertEquals("first", parser.getString());
+        assertEquals(START_ARRAY, parser.next());
+
+        assertEquals(2, parser.getArray().size());
+        assertEquals(END_ARRAY, parser.currentEvent());
+
+        assertEquals(KEY_NAME, parser.next());
+        assertEquals("second", parser.getString());
+        assertEquals(VALUE_STRING, parser.next());
+        assertEquals("after", parser.getString());
+        assertEquals(END_OBJECT, parser.next());
+        assertFalse(parser.hasNext());
+    }
+
+    @Test
+    void shouldLeaveScalarValuePositionUnchangedAfterGetValue()
+    {
+        JsonParser parser = parserFor("""
+            first: alpha
+            second: beta
+            """);
+
+        assertEquals(START_OBJECT, parser.next());
+        assertEquals(KEY_NAME, parser.next());
+        assertEquals(VALUE_STRING, parser.next());
+        assertEquals("alpha", ((jakarta.json.JsonString) parser.getValue()).getString());
+        assertEquals(VALUE_STRING, parser.currentEvent());
+
+        assertEquals(KEY_NAME, parser.next());
+        assertEquals("second", parser.getString());
+        assertEquals(VALUE_STRING, parser.next());
+        assertEquals("beta", parser.getString());
+        assertEquals(END_OBJECT, parser.next());
+        assertFalse(parser.hasNext());
     }
 
     @Test

